lib/Sema/CSGen.cpp

```diff
@@ -1176,9 +1176,18 @@ namespace {
 
     virtual Type visitCodeCompletionExpr(CodeCompletionExpr *E) {
       CS.Options |= ConstraintSystemFlags::SuppressDiagnostics;
-      return CS.createTypeVariable(CS.getConstraintLocator(E),
-                                   TVO_CanBindToLValue |
-                                   TVO_CanBindToNoEscape);
+      auto locator = CS.getConstraintLocator(E);
+      auto ty = CS.createTypeVariable(locator,
+                                      TVO_CanBindToLValue |
+                                      TVO_CanBindToNoEscape);
+
+      // Defaults to the type of the base expression.
+      if (auto base = E->getBase()) {
+        CS.addConstraint(ConstraintKind::Defaultable, ty, CS.getType(base),
+                         locator);
+      }
```

Comment:
Not sure this is desirable. But this is needed to preserve the current behavior in several test cases.

Reply:
Am I reading this correctly that we are saying that in
```
foo(something.<complete>)
```
the type of `something.completed_expr` must be defaultable to the type of `something`?  What kinds of tests changes do you see without this?

Reply:
`ConstraintKind::Defaultable` means "the first type *can be defaulted* to the second", not "must be defaultable", IIUC.
The failing test case was:
```
let val = Something.<HERE>
```
In this case `val ` is type checked to the type of `Something` so each result has type relation (e.g. `Invalid` for `Void` result). Without `Defaultable`, since `val` cannot be resolved, there's no expected type here. (I'm not saying that `val` as `Something` is the right thing)

Reply:
I see. That seems like an unfortunate way to say "non-Void".  What happens if there is an expected type and it really is Void?

Reply:
It seems like what we really want here is for code completion expression to be marked as a hole by default, just like we do in some other cases, that would solve a case where there is absolutely no contextual information available (`let _ = foo.<#complete#>` just like Rintaro mentioned).

I'm still thinking about this but it might be possible to generate constraints here in a way which would match what we do for `DefineMemberBasedOnUse` fix if member couldn't be found, that would guarantee that there is either a type or a hole inferred for a member reference.

Reply:
I think the least disruptive change would be to default code completion to `UnresolvedType` regardless of whether it's a member lookup or not, we also currently use `UnresolvedType` to indicate holes so it's fitting.